nntrainer/layers/lstmcell_core.cpp

```diff
@@ -0,0 +1,519 @@
+    next_hidden_state_derivative.reshape({batch_size, 1, 1, unit});
+  }
+
+  Tensor &cell_state = context.getTensor(wt_idx[2]);
+  cell_state.reshape({max_timestep, 1, batch_size, unit});
+  Tensor next_cell_state = cell_state.getBatchSlice(timestep, 1);
+  next_cell_state.reshape({batch_size, 1, 1, unit});
+  Tensor next_cell_state_derivative;
+  if (training) {
+    Tensor &cell_state_derivative = context.getTensorGrad(wt_idx[2]);
+    cell_state_derivative.reshape({max_timestep, 1, batch_size, unit});
+    next_cell_state_derivative =
+      cell_state_derivative.getBatchSlice(timestep, 1);
+    next_cell_state_derivative.reshape({batch_size, 1, 1, unit});
+  }
+
+  outputs[0] = Var_Grad(next_hidden_state, next_hidden_state_derivative,
+                        context.getTensorName(wt_idx[1]));
+  outputs[1] = Var_Grad(next_cell_state, next_cell_state_derivative,
+                        context.getTensorName(wt_idx[2]));
+}
+
+const std::vector<Var_Grad *> getOutputs(std::vector<Var_Grad> &outputs) {
+  std::vector<Var_Grad *> ret(outputs.size());
+  for (unsigned int i = 0; i < outputs.size(); ++i) {
+    ret[i] = &outputs[i];
+  }
+  return ret;
+}
+
+void fillTensors(std::vector<Var_Grad> &tensors, RunLayerContext &context,
+                 bool training, const std::vector<unsigned int> &wt_idx,
+                 const unsigned int max_timestep, const unsigned int timestep) {
+  tensors.resize(1);
+
+#if ENABLE_SHARING_WT_IDX
+  const Tensor &ifgo = context.getTensor(wt_idx[0]);
+  Tensor empty;
+  const Tensor &ifgo_derivative =
+    training ? context.getTensorGrad(wt_idx[0]) : empty;
+  tensors[0] =
+    Var_Grad(ifgo, ifgo_derivative, context.getTensorName(wt_idx[0]));
+#else
+  const TensorDim &output_dim = context.getOutput(0).getDim();
+  const unsigned int batch_size = output_dim.batch();
+  const unsigned int unit = output_dim.width();
+
+  Tensor &ifgo = context.getTensor(wt_idx[0]);
+  const unsigned int NUM_GATE = ifgo.width() / unit;
+  ifgo.reshape({max_timestep, 1, batch_size, NUM_GATE * unit});
+  Tensor ifgo_t = ifgo.getBatchSlice(timestep, 1);
+  ifgo_t.reshape({batch_size, 1, 1, NUM_GATE * unit});
+  Tensor ifgo_derivative_t;
+  if (training) {
+    Tensor &ifgo_derivative = context.getTensorGrad(wt_idx[0]);
+    ifgo_derivative.reshape({max_timestep, 1, batch_size, NUM_GATE * unit});
+    ifgo_derivative_t = ifgo_derivative.getBatchSlice(timestep, 1);
+    ifgo_derivative_t.reshape({batch_size, 1, 1, NUM_GATE * unit});
+  }
+  tensors[0] =
+    Var_Grad(ifgo_t, ifgo_derivative_t, context.getTensorName(wt_idx[0]));
+#endif
+}
+
+const std::vector<Var_Grad *> getTensors(std::vector<Var_Grad> &tensors) {
+  std::vector<Var_Grad *> ret(tensors.size());
+  for (unsigned int i = 0; i < tensors.size(); ++i) {
+    ret[i] = &tensors[i];
+  }
+  return ret;
+}
+
+enum INDEX {
+  INPUT,
+  HIDDEN_STATE_IN,
+  CELL_STATE_IN,
+  HIDDEN_STATE_OUT = 0,
+  CELL_STATE_OUT
+};
```

Comment:
important: this will be interpreted as

Input = 0
Hidden_state_in = 1
cell_state_in = 2
hidden_state_out = 0
cell_state_out = 1

highly likely that this is causing problem somewhere. If not, we need to check why this is not causing any prob as well.